torch/csrc/api/include/torch/nn/module.h

```diff
@@ -0,0 +1,130 @@
+#pragma once
+
+#include <torch/tensor.h>
+
+#include <memory>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+namespace at {
+enum class ScalarType;
+} // namespace at
+
+namespace torch { namespace nn {
+
+struct Archive;
+struct BufferCursor;
+struct ModuleCursor;
+struct ParameterCursor;
+
+/// The base class for all torch modules.
+class Module {
+ public:
+  /// Tells the base `Module` about the name of the submodule.
+  explicit Module(std::string name);
+  virtual ~Module();
+
+  /// Returns the name of the `Module`.
+  const std::string& name() const noexcept;
+
+  /// Performs a recursive clone of the entire module hierarchy. This is to
+  /// enable deep copying in a polymorphic setting, i.e. when all you have is a
+  /// `Module*` but want to clone the whole module hierarchy.
+  virtual std::unique_ptr<Module> clone();
+
+  /// Takes a list of input variables and computes a list of output variables.
+  virtual std::vector<Tensor> forward(const std::vector<Tensor>& inputs) = 0;
```

Comment:
We should also have `operator()` (not virtual), so the interface will match what we have in Python. Right now this requires people to access `.forward`, which is discouraged.